### DISCUSSION TITLE
Multithreading on remote training?

### DISCUSSION THREAD

**Discussion by mollywestbrook** [talmolab/sleap#1395] [Help!] [answered]
Quick question. I'm planning to use my university's computer cluster for some remote training, and there is no `-t` command in `sleap-train`. Is there a way to specify the number of threads for the command, or does sleap operate on a single thread if run through the CPU? Alternatively, if we just specify to use GPU instead of CPU will we not have to worry about this?

**Comment by talmo** [comment, marked as answer]
Hi @mollywestbrook,

Right so, SLEAP uses TensorFlow for its neural network backend, and specifically the [`tf.data`](https://www.tensorflow.org/guide/data) API for generating data during training. This automatically parallelizes all of the data processing operations (including multiprocessing and multithreading) based on its internal optimization algorithms. This means that it'll use as many threads/processes as needed to maximize throughput, so there's no need to specify how many threads to use. It usually just uses the max available threads in the environment it's running in.

Data pipelines are executed on the CPU while neural network optimization is done all on the GPU. All GPU operations are inherently parallel, so there's not much to be tuned there.

Let us know if you have any questions!

Talmo

**Reply by mollywestbrook** [reply]
Thanks for the helpful feedback! At this point we are up and running on our cluster, and have skyrocketed from 4 iterations/sec to ~380 iterations/sec. For anyone else who is interested, our slurm scripts for top-down training: [centered-instance](https://github.com/mollywestbrook/mollys-sleap-code/blob/main/train-sleap-centered-instance.sh) and [centroid](https://github.com/mollywestbrook/mollys-sleap-code/blob/main/train-sleap-centroid.sh).
